pkg/operator/operator_test.go

```diff
@@ -0,0 +1,169 @@
+				configv1.OperatorProgressing: configv1.ConditionFalse,
+				configv1.OperatorFailing:     configv1.ConditionTrue,
+			},
+		},
+		{
+			platform:          bareMetalPlatform,
+			expectedSyncError: true,
+			expectedConditions: map[configv1.ClusterStatusConditionType]configv1.ConditionStatus{
+				configv1.OperatorAvailable:   configv1.ConditionTrue,
+				configv1.OperatorProgressing: configv1.ConditionFalse,
+				configv1.OperatorFailing:     configv1.ConditionTrue,
+			},
+		},
+		{
+			platform:          kubemarkPlatform,
+			expectedSyncError: true,
+			expectedConditions: map[configv1.ClusterStatusConditionType]configv1.ConditionStatus{
+				configv1.OperatorAvailable:   configv1.ConditionTrue,
+				configv1.OperatorProgressing: configv1.ConditionFalse,
+				configv1.OperatorFailing:     configv1.ConditionTrue,
+			},
+		},
+		{
+			platform: configv1.VSpherePlatform,
+			expectedConditions: map[configv1.ClusterStatusConditionType]configv1.ConditionStatus{
+				configv1.OperatorAvailable:   configv1.ConditionTrue,
+				configv1.OperatorProgressing: configv1.ConditionFalse,
+				configv1.OperatorFailing:     configv1.ConditionFalse,
+			},
+		},
+		{
+			platform: configv1.NonePlatform,
+			expectedConditions: map[configv1.ClusterStatusConditionType]configv1.ConditionStatus{
+				configv1.OperatorAvailable:   configv1.ConditionTrue,
+				configv1.OperatorProgressing: configv1.ConditionFalse,
+				configv1.OperatorFailing:     configv1.ConditionFalse,
+			},
+		},
+		{
+			platform: "bad-platform",
+			expectedConditions: map[configv1.ClusterStatusConditionType]configv1.ConditionStatus{
+				configv1.OperatorAvailable:   configv1.ConditionTrue,
+				configv1.OperatorProgressing: configv1.ConditionFalse,
+				configv1.OperatorFailing:     configv1.ConditionFalse,
+			},
+		},
+	}
+	for _, tc := range cases {
+		t.Run(string(tc.platform), func(t *testing.T) {
+			tempDir, err := ioutil.TempDir("", "TestOperatorSync")
+			if err != nil {
+				t.Fatalf("could not create the temp dir: %v", err)
+			}
+			defer os.RemoveAll(tempDir)
+
+			images := Images{
+				MachineAPIOperator: "test-mao",
+			}
+			imagesAsJSON, err := json.Marshal(images)
+			if err != nil {
+				t.Fatalf("failed to marshal images: %v", err)
+			}
+			imagesFilePath := filepath.Join(tempDir, "test-images.json")
+			if err := ioutil.WriteFile(imagesFilePath, imagesAsJSON, 0666); err != nil {
+				t.Fatalf("could not write the images file: %v", err)
+			}
+
+			infra := &configv1.Infrastructure{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "cluster",
+				},
+				Status: configv1.InfrastructureStatus{
+					Platform: tc.platform,
+				},
+			}
+
+			optr := Operator{
+				eventRecorder: record.NewFakeRecorder(5),
+				osClient:      fakeconfigclientset.NewSimpleClientset(infra),
+				imagesFile:    imagesFilePath,
+			}
+
+			err = optr.sync("test-key")
+			if tc.expectedSyncError {
+				if !assert.Error(t, err, "unexpected sync success") {
+					t.Fatal()
```

Comment:
why `Fatal`? from the docs: "it will call `failNow` and stops function execution", why not  just asset.Error so the loop keep going and we output for each case
cc @frobware

Reply:
With sub-tests, the call to `Fatal` will only fail the one sub-test. The rest of the sub-tests will still execute.

Reply:
so what's the different over just `if !tc.expectedNoop { assert.Error(t, err, "unexpected sync success")}`

Reply:
The difference is that with your code, the rest of the sub-test will be executed. So there will be more test errors later on in the sub-test when it comes to testing the conditions set on the clusteroperator. That is not terribly; it does make the test results harder to analyze due to more noise, however.

I am fine with removing the call to `Fatal` here, if you prefer. The call to `Fatal` at https://github.com/openshift/machine-api-operator/blob/67a8e84e4a5b7105e75f6d0cb4db777c61e3231d/pkg/operator/operator_test.go#L108-L111 needs to stay to prevent a panic when the clusteroperator does not exist.